iron-input.html

```diff
@@ -125,11 +169,52 @@
      * @suppress {checkTypes}
      */
     _bindValueChanged: function() {
+      if (typeof this.bindValue !== 'string') {
+        // it will trigger another _bindValueChanged event
+        this.bindValue = !(this.bindValue || this.bindValue === 0) ? '' : this.bindValue + '';
+      } else {
+        this._updateValueIfNeeded();
+        // manually notify because we don't want to notify until after setting value
+        this.fire('bind-value-changed', {value: this.bindValue});
+      }
+    },
+
+    _bindValueAsNumberChanged: function() {
+      // ensure it is a number
+      if (typeof this.bindValueAsNumber !== 'number') {
+        // it will trigger another _bindValueAsNumberChanged event
+        if (isNaN(this.bindValueAsNumber)) {
+          this.bindValueAsNumber = NaN;
+        } else {
+          this.bindValueAsNumber = parseFloat(this.bindValueAsNumber);
+        }
+      } else {
+        this._updateValueAsNumberIfNeeded();
+        // manually notify because we don't want to notify until after setting valueAsNumber
+        this.fire('bind-value-as-number-changed', {value: this.bindValueAsNumber});
+      }
+    },
+
+    _updateValueIfNeeded: function() {
       if (this.value !== this.bindValue) {
-        this.value = !(this.bindValue || this.bindValue === 0) ? '' : this.bindValue;
+        this.value = this.bindValue;
+        this._updateBindValues();
+      }
+    },
+
+    _updateValueAsNumberIfNeeded: function(){
+      // NaN is != from NaN, must check if at least one of them is a number!
+      if ((this.type === 'number' || this.type === 'date') && this.valueAsNumber !== this.bindValueAsNumber && (this.valueAsNumber === this.valueAsNumber || this.bindValueAsNumber === this.bindValueAsNumber)) {
```

Comment:
It might be worth polyfilling [`Number.isNaN`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN). It has the benefit for more [straight-forward coercion semantics](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN), and seems to have [competitive or superior performance](https://jsperf.com/number-isnan-vs-handrolled) to other strategies where implemented.